include/openssl/evp.h

```diff
@@ -690,20 +690,53 @@ OPENSSL_EXPORT int EVP_PKEY_keygen_init(EVP_PKEY_CTX *ctx);
 // containing the result. It returns one on success or zero on error.
 OPENSSL_EXPORT int EVP_PKEY_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY **out_pkey);
 
-OPENSSL_EXPORT int EVP_PKEY_encapsulate_init(EVP_PKEY_CTX *ctx,
-                                             const OSSL_PARAM params[]);
-OPENSSL_EXPORT int EVP_PKEY_encapsulate(EVP_PKEY_CTX *ctx,
-                                        unsigned char *wrapped_key,
-                                        size_t *wrapped_key_len,
-                                        unsigned char *gen_key,
-                                        size_t *gen_key_len);
-OPENSSL_EXPORT int EVP_PKEY_decapsulate_init(EVP_PKEY_CTX *ctx,
-                                             const OSSL_PARAM params[]);
-OPENSSL_EXPORT int EVP_PKEY_decapsulate(EVP_PKEY_CTX *ctx,
-                                        unsigned char *unwrapped,
-                                        size_t *unwrapped_len,
-                                        const unsigned char *wrapped,
-                                        size_t wrapped_len);
+// EVP_PKEY_encapsulate is an operation defined for a KEM (Key Encapsulation
+// Mechanism). For the KEM specified in |ctx|, the function:
+//   1. generates a random value and writes it to |shared_secret|,
+//   2. encapsulates the shared secret, producing the ciphertext, by using
+//      the public key in |ctx|, and writes the ciphertext to |ciphertext|,
+//   3. writes the length of |ciphertext| and |shared_secret| to
+//      |ciphertext_len| and |shared_secret_len|.
+//
+// If the given |ciphertext| is NULL it is assumed that the caller is doing
+// a size check: the function will write the size of the ciphertext and the
+// shared secret in |ciphertext_len| and |shared_secret_len| and return 1.
+// If |ciphertext| is non-NULL it is assumed that the caller is performing
+// the actual operation, so it is checked if the lengths of the output buffers,
+// |ciphertext_len| and |shared_secret_len|, are large enough for the KEM.
+//
+// NOTE: no allocation is done in the function, the caller is expected to
+// provide large enough |ciphertext| and |shared_secret| buffers.
+//
+// It returns one on success or zero on error.
+OPENSSL_EXPORT int EVP_PKEY_encapsulate(EVP_PKEY_CTX *ctx          /* IN  */,
+                                        uint8_t *ciphertext        /* OUT */,
+                                        size_t  *ciphertext_len    /* OUT */,
+                                        uint8_t *shared_secret     /* OUT */,
+                                        size_t  *shared_secret_len /* OUT */);
+
+// EVP_PKEY_decapsulate is an operation defined for a KEM (Key Encapsulation
+// Mechanism). For the KEM specified in |ctx|, the function:
+//   1. decapsulates the shared secret from the given |ciphertext| using the
+//      secret key given within |ctx| and writes it to |shared_secret|,
```

Comment:
Just a nit suggestion:
```suggestion
//      secret key configured in |ctx| and writes it to |shared_secret|,
```